src/world/conquest_system.cpp

```diff
@@ -20,3 +20,342 @@ along with this program.  If not, see http://www.gnu.org/licenses/
 */
 
 #include "conquest_system.h"
+
+/*
+===========================================================================
+
+Copyright (c) 2010-2015 Darkstar Dev Teams
+
+                        This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+                                         (at your option) any later version.
+
+                                         This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#include "message_server.h"
+
+ConquestSystem::ConquestSystem(std::shared_ptr<SqlConnection> sql)
+: zmqSql(sql)
+{
+}
+
+bool ConquestSystem::handleMessage(MSGSERVTYPE     type,
+                                   zmq::message_t* extra,
+                                   zmq::message_t* packet,
+                                   zmq::message_t* from,
+                                   in_addr         from_addr,
+                                   uint16          from_port)
+{
+    if (type != MSGSERVTYPE::MSG_MAP2WORLD_REGIONAL_EVENT)
```

Comment:
how would you register these? I was thinking of a map of tuple<type, subtype> but the syntax was kinda ugly

Reply:
You're quite close:

In this block:
```cpp
    // Setup message handlers
    std::vector<std::shared_ptr<MessageHandler>> handlers;
    handlers.push_back(conquestSystem);
    messageServer = std::make_unique<message_server_wrapper_t>(zmqSql, handlers, std::ref(m_RequestExit));
```

You've already created the conquest system, just create a map of the defining message type to a shared_ptr to the ConquestSystem. The map will be of pointers to MessageHandlers, since it's an interface that's all fine.

Reply:
Yeah, instead of this long chain of fallthroughs, you should be mapping things to function pointers

Reply:
the part that i'd like to make prettier is mapping the type and subtype to a handler. Few options:
- using a tuple<type, subtype> as key for the handler map
- using a nested map<type, map<subtype, handler>>
- using a uint16 key and just use key = type | (subtype << 8)

I am leaning towards the latter

Reply:
You have 2x sets of keys, there's no need to combine them or do a single dispatch.

You do one lookup of destination key to shared ptr value of the relevant message handler in the message server, then forward the payload.

You do the next lookup of subject (message type) once you're in the correct handler, mapping the key to a function pointer of the function that'll handle the payload

Reply:
Type:  REGIONALMSGTYPE : uint8
Subtype: CONQUESTMSGTYPE : uint8

Reply:
If each messageHandler ends up being on its own thread, then they'll have to communicate using locks or with a lockless queue, also

Reply:
Since we don't know our requirements for all handlers yet, I was going to just keep the same arguments to the message parse method (`zmq::message_t* extra, zmq::message_t* packet, from_ip, from_port`). Would there be issues with that?

Reply:
Actually packet seems more meant for the request / reply mechanism so I think I can do with just the data vector as you suggested, and from_ip / port

Reply:
Remember (so far, it might change in the future) that the idea is to not generate actual packet data for the client in the world server.

It's a little annoying and inconvenient that they use the same structures and mechanisms, but ultimately, world server doesn't have all the same information that map does - nor does it matter.

Reply:
so if we send around an std::vector<uint8>, is this the way we would access properties?
```
        int32 points;
        std::memcpy(&points, payload.data(), sizeof(int32));
```

Any other options?

Reply:
What do you mean by properties? For a given message type, you know what type of item is going in, vs what is coming out?
```
std::memcpy(&points, payload.data(), sizeof(int32));
```
Should work fine, `.data()` can be abused for storage of whatever you want - given that it's contiguous memory. If you want to store something more complex, it's defined as a vector of uint8's (bytes). So you can just construct the payload as sizeof(your_struct) and then write/read out the data from `.data()`

Reply:
If you _really really_ wanted to, you could also start using flatbuffers, but that feels overkill for this